tools/deploy_firmware.py

```diff
@@ -0,0 +1,287 @@
+#!/usr/bin/env python3
+
+import time
+import sys
+import os
+import subprocess
+import difflib
+import argparse
+import shutil
+from typing import List
+
+import pexpect
+import pexpect.fdpexpect
+import pexpect.pxssh
+import serial
+
+
+class PrplwrtDevice:
+    """Represents a prplWrt device.
+
+    Offers methods to check if a device needs to be upgraded and to do the actual upgrade.
+
+    It needs to have access to the artifacts of a build job to determine when an upgrade
+    is needed (see `artifacts_dir`).
+    """
+
+    serial_prompt = "_pexpect_prompt_ "
+    """For serial connections we will set this prompt to make it easier to "expect" it."""
+
+    baudrate = 115200
+    """The baudrate of the serial connection to the device."""
+
+    def __init__(self, device: str, name: str, image: str, username: str = "root"):
+        """
+
+        Parameters
+        -----------
+        device: str
+            The name of the platform (example: netgear-rax40).
+        name: str
+            The name of the device (it should ne reachable through ssh without a password).
+        image: str
+            The name of the image that can be used to upgrade the device.
+        username: str, optional
+            The username to use when connecting to the device over SSH.
+        """
+        self.device = device
```

Comment:
nitpick: Consider not making the device name an argument of the constructor but rather a property of the device class, overwritten by each subclass (the base class implementation can raise `NotImplementedError`)